data/data-pipeline/data_pipeline/comparison_tool/src/utils.py

```diff
@@ -0,0 +1,330 @@
+        population_column: the column that includes population count per tract
+        geoid_column: the geoid10_tract column
+
+    Returns:
+        population: the high-level overview df
+    """
+    population = joined_frame.groupby([comparator_column, score_column]).agg(
+        {population_column: ["sum"], geoid_column: ["count"]}
+    )
+
+    population["share_of_tracts"] = (
+        population[geoid_column] / population[geoid_column].sum()
+    )
+
+    population["share_of_population_in_tracts"] = (
+        population[population_column] / population[population_column].sum()
+    )
+
+    population.columns = [
+        "Population",
+        "Count of tracts",
+        "Share of tracts",
+        "Share of population",
+    ]
+    return population
+
+
+def get_demo_series(
+    grouping_column: str,
+    joined_frame: pd.DataFrame,
+    demo_columns: list,
+) -> pd.DataFrame:
+    """Helper function to produce demographic information"""
+    return (
+        joined_frame[joined_frame[grouping_column]][demo_columns]
+        .mean()
+        .T.rename(grouping_column)
+    )
+
+
+def get_tract_level_grouping(
+    joined_frame: pd.DataFrame,
+    score_column: str,
+    comparator_column: str,
+    demo_columns: list,
+) -> pd.DataFrame:
+    """Function to produce segmented statistics (tract level)"""
+    group_list = [score_column, comparator_column]
+    grouping_frame = (
+        joined_frame[joined_frame[group_list].sum(axis=1) > 0]
+        .groupby(group_list)[demo_columns]
+        .mean()
+        .reset_index()
+    )
+    grouping_frame[score_column] = grouping_frame[score_column].map(
+        {True: "CEJST", False: "Not CEJST"}
+    )
+    grouping_frame[comparator_column] = grouping_frame[comparator_column].map(
+        {True: "Comparator", False: "Not Comparator"}
+    )
+    return grouping_frame.set_index([score_column, comparator_column]).T
+
+
+def format_multi_index_for_excel(
+    df: pd.DataFrame, rename_str: str = "Variable"
+) -> pd.DataFrame:
+    """Helper function for multiindex printing"""
+    df = df.reset_index()
+    df.columns = [rename_str] + [
+        ", ".join(col_tuple).strip()
+        for col_tuple in df.columns[1:].to_flat_index()
+    ]
+    return df
+
+
+def get_final_summary_info(
+    population: pd.DataFrame,
+    comparator_file: str,
+    geoid_col: str,
+) -> tuple[pd.DataFrame, str]:
+    also_cejst = population.loc[(True, True)] / population.loc[(True,)].sum()
+
+    states_represented = (
+        pd.read_csv(
+            comparator_file, usecols=[geoid_col], dtype={geoid_col: str}
+        )[geoid_col]
+        .str[:2]
+        .unique()
+    )
+    states = ", ".join(
+        [
+            FIPS_MAP[state]
+            if (state in FIPS_MAP)
+            else f"territory (fips {state})"
+            for state in states_represented
+        ]
+    )
+    return also_cejst, states
+
+
+def construct_weighted_statistics(
+    _joined_frame: pd.DataFrame,
```

Comment:
nit - usually an underscore in Python indicates private use (https://dbader.org/blog/meaning-of-underscores-in-python#:~:text=The%20underscore%20prefix%20is%20meant,public%E2%80%9D%20variables%20like%20Java%20does.). Is this what you meant here?

Reply:
good catch! nope, i just meant it gets copied